Implement ticket activity page

song50119 opened the pull request:
PR implement for ticket page activity
<img width="795" alt="Screen Shot 2021-05-27 at 01 00 04" src="https://user-images.githubusercontent.com/9800579/119708970-e150dc00-be86-11eb-8cfa-b1629068c1bb.png">

oshorefueled commented:
The history page seems to work when using one wallet. However, it's not fully functional with multiple wallets. First, how multiple wallets are displayed. A dropdown would be more suitable for displaying users' wallets. You can use the "receive" page as a reference for that. 

<img width="969" alt="Screenshot 2019-12-15 at 1 55 53 AM" src="https://user-images.githubusercontent.com/15804688/70856491-5d976f80-1ede-11ea-9b3b-5413f2015f04.png">

I get an empty transactions result when I try switching between wallets and that's just one of the issues. Kindly create multiple wallets and test.

<img width="970" alt="Screenshot 2019-12-15 at 2 09 57 AM" src="https://user-images.githubusercontent.com/15804688/70856622-56716100-1ee0-11ea-9805-21500b2d1990.png">

oshorefueled commented on terminal/pages/history.go:
```diff
@@ -20,7 +20,7 @@ const (
 )
 
-const txPerPage int32 = 25
+const txPerPage int32 = 10
```
You're working on the history page for fyne. You shouldn't make changes to other files that don't affect the feature you're adding. If there's a need to clean up code or make modifications to unrelated parts of the code base, that should be done on another pull request.  Please fix other parts of this PR that this applies to.

Sirmorrison replied:
Please you may need to be more explicit as to the pages.

oshorefueled commented on cli/go.sum:
```diff
@@ -14,6 +14,7 @@ github.com/agl/ed25519 v0.0.0-20170116200512-5312a6153412/go.mod h1:WPjqKcmVOxf0
 github.com/akavel/rsrc v0.8.0/go.mod h1:uLoCtb9J+EyAqh+26kdrTgmzRBFPGOolLWKpdxkKq+c=
 github.com/asdine/storm v0.0.0-20190216191021-fe89819f6282 h1:DmSVc81daQAPvXwcCZi0W6A14sTCYQ1QI21C0E37KoY=
 github.com/asdine/storm v0.0.0-20190216191021-fe89819f6282/go.mod h1:cMLKpjHSP4q0P133fV15ojQgwWWB2IMv+hrFsmBF/wI=
+github.com/atotto/clipboard v0.1.2 h1:YZCtFu5Ie8qX2VmVTBnrqLSiU9XOWwqNRmdT3gIQzbY=
```
unrelated file to the pull request, this file shouldn't be part of the PR. When working on a feature, you should make modifications to only areas that affect or are related to what you're working on.

oshorefueled replied:
> unrelated file to the pull request, this file shouldn't be part of the PR. When working on a feature, you should make modifications to only areas that affect or are related to what you're working on.

This comment also applies to the files where you applied "go fmt" like the transactionList code. You also made modifications to the receives page code. If a code file doesn't relate to the feature you're working on, try not to make modifications to it. If you feel any area needs modification or improvement, you can bring up an issue and another PR can be created for that. Merge conflicts would be avoided this way. Kindly remove unrelated files to this PR.

oshorefueled commented on fyne/helpers/amount.go:
```diff
@@ -0,0 +1,59 @@
+package helpers
```
none of this code seems to be used anywhere, was it previously used somewhere?

Sirmorrison replied:
it came in after rebase. will remove it for now.

oshorefueled commented:
Please remove this file https://github.com/raedahgroup/godcr/blob/28a531acf89b55f9ab87a2e920240d443c6cef9b/cmd/godcr-fyne/.~bash-profile.swp from your PR

metaclips commented on fyne/pages/history.go:
```diff
@@ -1,10 +1,681 @@
+	if txHistory.allTxCount > int(txHistory.TotalTxFetched) {
+		if txHistory.txTable.Container.Offset.Y == 0 {
+			if txTableRowCount < int(txPerPage) {
+				return
```
If I am right this checks if the scroller is at the starting point and then if the amount of transactions in the table is less than the min tx per scroll which is 25 it returns, this shouldn't be because if it's a new wallet and there is an incoming transactions the table won't be updated.

metaclips commented on fyne/pages/history.go:
```diff
@@ -1,10 +1,681 @@
+			})
+		} else if scrollPosition < 0.5 {
+			if txHistory.TotalTxFetched == txPerPage {
+				time.AfterFunc(time.Second*4, func() {
```
You shouldn't update every 4 seconds, on a fast scroll <0.5 parameter might not be met, also I really don't see why we are calling a time.AfterFunc recursively, for every time users switch to history page a new time.AfterFunc call is made this can better be passed here https://github.com/raedahgroup/godcr/blob/0dc5f3cac669f276f1244f061976bdcf17212380/fyne/pages/display.go#L88 making sure to return function is history page is in view

metaclips commented on fyne/pages/history.go:
```diff
@@ -1,10 +1,681 @@
+
+				time.AfterFunc(time.Second*2, func() {
+					if tabMenu.CurrentTabIndex() == 1 {
+						messageLabel.SetText("")
```
This should be messageLabel.Hide()

Sirmorrison replied:
The idea was not to hide the message label. Reason is it distorts the page when ever the message displays.

metaclips replied:
While splitting components to handler package, place a variable of *widget.VBox in a struct that should house all component on history page then when you make changes, only the Box should be refreshed, this was done in send page and fixes issues with distortion.

Sirmorrison commented:
> You should try to split components to separate functions and place them in the handler package for easy review also scrolling seems very slow, can we reduce transaction table per scroll to 15 instead of 25?

As a matter of uniformity, across all godcr interfaces, the txperpage is 25.

metaclips commented:
> > You should try to split components to separate functions and place them in the handler package for easy review also scrolling seems very slow, can we reduce transaction table per scroll to 15 instead of 25?
> 
> As a matter of uniformity, across all godcr interfaces, the txperpage is 25.

Scrolling is slow and table contents distorts while scrolling we should reduce table contents.

oshorefueled commented:
Is it necessary that we use an horizontal scroll?
<img width="952" alt="Screenshot 2019-12-24 at 9 45 56 AM" src="https://user-images.githubusercontent.com/15804688/71404283-6dc3f300-2632-11ea-8125-e8657c6f12a0.png">

oshorefueled commented on cmd/go.mod:
```diff
@@ -6,10 +6,10 @@ require (
 	github.com/decred/slog v1.0.0
 	github.com/jessevdk/go-flags v1.4.0
 	github.com/jrick/logrotate v1.0.0
+	github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190928085114-bcc6e6b7769a // indirect
```
why do we need dcrlibwallet as a dependency here?

metaclips commented:
> Is it necessary that we use a horizontal scroll?
> <img alt="Screenshot 2019-12-24 at 9 45 56 AM" width="952" src="https://user-images.githubusercontent.com/15804688/71404283-6dc3f300-2632-11ea-8125-e8657c6f12a0.png">

There are contents we can do away with.
I don't see why the texts are needed with the icons in Direction line, type can also be removed and also hash text can be reduced so that the width wont be jumpy when users switch pages.

Sirmorrison commented:
> > Is it necessary that we use a horizontal scroll?
> > <img alt="Screenshot 2019-12-24 at 9 45 56 AM" width="952" src="https://user-images.githubusercontent.com/15804688/71404283-6dc3f300-2632-11ea-8125-e8657c6f12a0.png">
> 
> There are contents we can do away with.
> I don't see why the texts are needed with the icons in Direction line, type can also be removed and also hash text can be reduced so that the width wont be jumpy when users switch pages.

About the jumpy pages why dont we give the fyne interface a fixed min width?

Sirmorrison commented:
> Is it necessary that we use a horizontal scroll?
> <img alt="Screenshot 2019-12-24 at 9 45 56 AM" width="952" src="https://user-images.githubusercontent.com/15804688/71404283-6dc3f300-2632-11ea-8125-e8657c6f12a0.png">

resolved

oshorefueled commented:
The transaction details popup seems to be losing part of its content to the edges of the window.

<img width="744" alt="Screenshot 2019-12-27 at 10 31 14 AM" src="https://user-images.githubusercontent.com/15804688/71511855-17ee8580-2894-11ea-9455-a9e5a62acea8.png">

oshorefueled commented on cmd/go.mod:
```diff
@@ -9,7 +9,6 @@ require (
 	github.com/raedahgroup/godcr/app v0.0.0-00010101000000-000000000000
 	github.com/raedahgroup/godcr/cli v0.0.0-00010101000000-000000000000
 	github.com/raedahgroup/godcr/nuklear v0.0.0-00010101000000-000000000000
-	github.com/raedahgroup/godcr/terminal v0.0.0-00010101000000-000000000000 // indirect
```
Why was this line deleted?

oshorefueled commented on fyne/widgets/table.go:
```diff
@@ -11,67 +12,157 @@ type Table struct {
 }
 
+// Append widget adds to the bottom row of a table.
 func (table *Table) Append(data ...*widget.Box) {
```
You changed quite a number of code in the table.go. It's a widget I used on the overview page. It seemed to work well, any reason for the change? Would my existing implementation work without an issue if I used this?

Sirmorrison replied:
I had this these codes already before the overview page was implemented. Anyways have you tested your implementation on the overview page just to be sure there are no issues?

Sirmorrison commented:
> The transaction details popup seems to be losing part of its content to the edges of the window.
> 
> <img alt="Screenshot 2019-12-27 at 10 31 14 AM" width="744" src="https://user-images.githubusercontent.com/15804688/71511855-17ee8580-2894-11ea-9455-a9e5a62acea8.png">

so basically the issue is the @metaclips suggested that we reduce the width of the history page. this also affected the txdetails popup as its wider than the current history page.

Two things can done here. 
1) increase the width of the default history page to accommodate the details page.
2) reduce the width of the details popup and add a horizontal scroll to view the hidden details to the right. 
for now i will make it scroll horizontally.

Sirmorrison commented on fyne/pages/handler/historypagehandler/filterselector.go:
```diff
@@ -0,0 +1,133 @@
+
+	txCountForFilter, err := historyPage.MultiWallet.WalletWithID(historyPage.selectedWalletID).CountTransactions(allTxFilters["All"])
+	if err != nil {
+		historyPage.errorMessage = fmt.Sprintf("Cannot load history page page. Error getting transaction count for filter All: %s", err.Error())
```
this is not true is there is an error at any point the error is returned the app does not panic... you should do well to test this properly as i have done so from my end...

if you are changing `!=` to `==` you should comment out `err.Error()` else it will panic as there is no error to be returned. 

![image](https://user-images.githubusercontent.com/27733432/71560425-27bfc280-2a6a-11ea-9112-eb35eefbfd29.png)

metaclips replied:
<img width="479" alt="Screenshot 2019-12-30 at 19 58 06" src="https://user-images.githubusercontent.com/31141573/71596137-b69e0f00-2b3e-11ea-9b0c-a0477fb8d3dd.png">

This is what is shown when I make `err == nil` [here](https://github.com/Sirmorrison/godcr/blob/08c678bbe25a6b7609f1372848bae39567b60472/fyne/pages/handler/historypagehandler/txtablecomponents.go#L43), you are showing parts of history page components even when there was an error.
You shouldn't be passing errors to variables and then making a [return](https://github.com/Sirmorrison/godcr/blob/08c678bbe25a6b7609f1372848bae39567b60472/fyne/pages/handler/historypagehandler/filterselector.go#L34)

You are still calling up functions even if there was error on other functions called recently and then checking if any of the functions returned an error [here](https://github.com/Sirmorrison/godcr/blob/08c678bbe25a6b7609f1372848bae39567b60472/fyne/pages/handler/historypagehandler/inithistorypage.go#L81)

This shouldn't be as any of the functions that errors and returns early without initialising struct variable "HistoryPageData struct" could cause a panic in other functions if they use the uninitialised variable, you should instead pass return parameter "errors" to all functions that uses dcrlibwallet also InitHistoryPage function and then if any of the dcrlibwallet library functions return an error,
- Stop the failed function, return the error
- Stop history page initialisation in the handler package, return the error passed to InitHistoryPage
- Check if InitHistoryPage function which is to be called in history.go file returns an `err` if there was an error, only show on history page a label indicating that there was an error initialising history page.

Sirmorrison commented on fyne/pages/handler/historypagehandler/txtablecomponents.go:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	if prepend {
+		txTable.Prepend(txBox...)
```
This will function more like pagination reason is for every update a new table will be called and this will have only data from the last tx fetched.

Commit:
Implement ticket activity page
- Add ticket page activity
- Add initWalletDropDown to use on both ticket list and activity page
- Add filterTickets func to filter tickets and ticket activity
- Replace ticketActivityItemnInfo by ticketRowActivity to use on both page activity and ticket main page

oshorefueled commented on ui/tickets_activity_page.go:
```diff
@@ -0,0 +1,162 @@
+						return layout.Inset{Top: values.MarginPadding60}.Layout(gtx, func(gtx C) D {
+							return c.theme.Card().Layout(gtx, func(gtx C) D {
+								gtx.Constraints.Min = gtx.Constraints.Max
+								if pg.ticketTypeDropDown.SelectedIndex()-1 != -1 {
```
-1 in the if check is a magic number. Why is pg.ticketTypeDropDown being checked with "-1"

song50119 replied:
Dropdowns is init with item + 1
`c := &DropDown{
		isOpen:         false,
		items:          make([]DropDownItem, len(items)+1), 
...`

oshorefueled commented on ui/tickets_activity_page.go:
```diff
@@ -0,0 +1,162 @@
+	})
+}
+
+func (pg *ticketsActivityPage) dropDowns(gtx layout.Context) layout.Dimensions {
```
The wallet dropdown, sort dropdown and a page specific dropdown seems to be used in several pages. They could be implemented as "components" that can be called from each of those pages. This eliminates repetition.

oshorefueled commented on ui/compontents.go:
```diff
@@ -960,6 +960,92 @@ func ticketCard(gtx layout.Context, c pageCommon, t *wallet.Ticket) layout.Dimen
 }
 
+// ticketRowActivity layouts out ticket info, display ticket activities on the tickets_page and tickets_activity_page
+func ticketRowActivity(gtx layout.Context, c *pageCommon, t wallet.Ticket, index int) layout.Dimensions {
```
This looks like it should be named `ticketActivityRow`

Commit:
Add wallets dropdown to component
- Add createOrUpdateWalletDropDown and createOrderDropDown to component to be able to use on pages Ticket list, Tickets Activity and Transactions list
